cql3/statements/alter_keyspace_statement.cc

```diff
@@ -43,18 +44,17 @@ future<> cql3::statements::alter_keyspace_statement::check_access(query_processo
     return state.has_keyspace_access(_name, auth::permission::ALTER);
 }
 
-static bool validate_rf_difference(const std::string_view curr_rf, const std::string_view new_rf) {
-    auto to_number = [] (const std::string_view rf) {
-        int result;
-        // We assume the passed string view represents a valid decimal number,
-        // so we don't need the error code.
-        (void) std::from_chars(rf.begin(), rf.end(), result);
-        return result;
-    };
-
-    // We want to ensure that each DC's RF is going to change by at most 1
-    // because in that case the old and new quorums must overlap.
-    return std::abs(to_number(curr_rf) - to_number(new_rf)) <= 1;
+// We want to ensure that each DC's RF is going to change by at most 1
+static bool validate_rf_difference(const std::string& curr_rf, const std::string& new_rf) {
+    try {
+        return std::abs(std::stoi(curr_rf) - std::stoi(new_rf)) <= 1;
+    } catch (std::invalid_argument const& ex) {
+        on_internal_error(mylogger, format("validate_rf_difference expects integer arguments, "
```

Comment:
Is it indeed an internal error?
Can we get here if the user passed non-integer args as rf values?
If so, we should throw `exceptions::invalid_request_exception`

Reply:
> Is it indeed an internal error?

To the best of my knowledge, this should be considered an internal error, as we already have checks in place that'd reject anything else than natural numbers. The reason it wasn't working previously was because the code that called this function generated incorrect set of RFs (this set contained a non-numerical value). But with this bug fixed, the code here shouldn't receive anything else than natural numbers (i.e. >0).
See e.g.:
```
cqlsh> CREATE KEYSPACE ksm22 WITH REPLICATION = { 'class' : 'NetworkTopologyStrategy', 'replication_factor' : '-1' } AND TABLETS = {'initial': 2};
ConfigurationException: Replication factor must be numeric and non-negative, found '-1'
```
> Can we get here if the user passed non-integer args as rf values?

We have checks executed earlier that prevent accepting such malformed RFs, e.g.:
```
cqlsh> CREATE KEYSPACE ksm22 WITH REPLICATION = { 'class' : 'NetworkTopologyStrategy', 'replication_factor' : 'Benny' } AND TABLETS = {'initial': 2};
ConfigurationException: Replication factor must be numeric and non-negative, found 'Benny'
```
(plus also the check I posted to answer your previous part of this question).

> If so, we should throw exceptions::invalid_request_exception

So for the reasons above I made it an internal_error. Let me know if that's OK or you still want this to throw `exceptions::invalid_request_exception`

Reply:
> the code here shouldn't receive anything else than natural numbers

Make it a precondition of the function i.e. write a comment.

Better yet, change the type of the function to receive integers. The place where we validate that we're dealing with natural numbers? That same place should parse the strings into integers and we should be dealing with integers after that.

https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/